pkg/ratelimit/limitware/middleware.go

```diff
@@ -0,0 +1,216 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &Middleware{
+		store:      s,
+		keyFunc:    f,
+		meter:      meter,
+		reqCounter: rc,
+	}, nil
+}
+
+// Handle returns the HTTP handler as a middleware. This handler calls Take() on
+// the store and sets the common rate limiting headers. If the take is
+// successful, the remaining middleware is called. If take is unsuccessful, the
+// middleware chain is halted and the function renders a 429 to the caller with
+// metadata about when it's safe to retry.
+func (m *Middleware) Handle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Call the key function - if this fails, it's an internal server error.
+		key, err := m.keyFunc(r)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+
+		// Take from the store.
+		limit, remaining, reset, ok := m.store.Take(key)
+		resetTime := time.Unix(0, int64(reset)).UTC().Format(time.RFC1123)
+
+		// Set headers (we do this regardless of whether the request is permitted).
+		w.Header().Set(HeaderRateLimitLimit, strconv.FormatUint(limit, 10))
+		w.Header().Set(HeaderRateLimitRemaining, strconv.FormatUint(remaining, 10))
+		w.Header().Set(HeaderRateLimitReset, resetTime)
+
+		// Record request status
+		m.meter.RecordBatch(
+			r.Context(),
+			[]kv.KeyValue{kv.Bool("ok", ok)},
+			m.reqCounter.Measurement(1),
+		)
+
+		// Fail if there were no tokens remaining.
+		if !ok {
+			w.Header().Set(HeaderRetryAfter, resetTime)
+			http.Error(w, http.StatusText(http.StatusTooManyRequests), http.StatusTooManyRequests)
+			return
+		}
+
+		// If we got this far, we're allowed to continue, so call the next middleware
+		// in the stack to continue processing.
+		next.ServeHTTP(w, r)
+	})
+}
+
+// APIKeyFunc returns a default key function for ratelimiting on our API key header.
+func APIKeyFunc(ctx context.Context, scope string, db *database.Database) KeyFunc {
+	logger := logging.FromContext(ctx).Named("ratelimit")
+
+	return func(r *http.Request) (string, error) {
+		// Procss the API key
+		v := r.Header.Get("X-API-Key")
+		if v != "" {
+			// v2 API keys encode the realm
+			_, realmID, err := db.VerifyAPIKeySignature(v)
+			if err == nil {
+				logger.Debugw("limiting by api key v2 realm", "realm", realmID)
+				dig := sha1.Sum([]byte(fmt.Sprintf("%d", realmID)))
+				return fmt.Sprintf("apiserver:realm:%x", dig), nil
+			}
+
+			// v1 API keys do not, fallback to the database
+			app, err := db.FindAuthorizedAppByAPIKey(v)
+			if err == nil && app != nil {
+				logger.Debugw("limiting by api key v1 realm", "realm", app.RealmID)
+				dig := sha1.Sum([]byte(fmt.Sprintf("%d", app.RealmID)))
+				return fmt.Sprintf("%s:realm:%x", scope, dig), nil
+			}
+		}
+
+		// Get the remote addr
```

Comment:
This logic is duplicated - we should consider moving it into its own function for readability.

Reply:
oh fair, deduped.